app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,208 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HitRootIO.cc
+//---------------------------------------------------------------------------//
+#include "HitRootIO.hh"
+
+#include <regex>
+#include <G4AutoDelete.hh>
+#include <G4Event.hh>
+#include <G4RunManager.hh>
+#include <G4Threading.hh>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TObject.h>
+#include <TSystem.h>
+#include <TTree.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/io/Logger.hh"
+#include "accel/ExceptionConverter.hh"
+#include "accel/SetupOptions.hh"
+
+#include "GlobalSetup.hh"
+#include "SensitiveHit.hh"
+
+namespace
+{
+// Mutex within the HitRootIO scope
+G4Mutex HitRootIOMutex = G4MUTEX_INITIALIZER;
+}  // namespace
+
+namespace demo_geant
+{
+G4ThreadLocal HitRootIO* HitRootIO::instance_ = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a ROOT output file for each worker except the master thread if MT
+ */
+HitRootIO::HitRootIO()
+{
+    G4AutoLock lock(&HitRootIOMutex);
+    gSystem->Load("libHitClassesDict");
+
+    file_name_ = std::regex_replace(
+        GlobalSetup::Instance()->GetSetupOptions()->output_file,
+        std::regex("json"),
+        "root");
+
+    if (G4Threading::IsWorkerThread())
+    {
+        file_name_ += std::to_string(G4Threading::G4GetThreadId());
+    }
+
+    if (G4Threading::IsWorkerThread()
+        || !G4Threading::IsMultithreadedApplication())
+    {
+        file_ = TFile::Open(file_name_.c_str(), "recreate");
+        CELER_VALIDATE(file_->IsOpen(), << "Failed to open " << file_name_);
+        tree_ = new TTree(
+            "Events", "Hit collections", this->SplitLevel(), file_);
+        G4AutoDelete::Register(tree_);
```

Comment:
No problem to make it unique_ptr here.  Any suggestion how to change  1) `tree_->Branch(...)` 2) `tree_->Fill() `and 3) `file_ = tree_->GetCurrentFile()`

Reply:
See `celeritas/io/RootFileManager` (though that uses a unique pointer with a custom destructor): you should just be able to have a `std::unique_ptr<TTree> tree_` and `tree_.reset(new TTree(...))` or `tree_ = std::make_unique<TTree>(....)`.

Reply:
@sethrj I tried to make them unique_ptr (with both RootUniquePtr and the plain std::unique_ptr), which compiles okay and runs successfully up to the file merging, but somehow hits a segmentation fault at the end of program due to a conflict between `G4ThreadLocalSingletion<demo_geant::HitRootIO>::~G4ThreadLocalSingleton()` and `~unique_ptr()`.  So, it looks that this should go together with removing G4ThreadLocalSingleton from the code.  Can you take over this and refactor the code as your way (i.e., using the ActionInitialization interface)?

Reply:
Did you apply the `unique_ptr` only to the `TTree` or both the `TFile` and `TTree`?  Either way, did you 'insure' that the `TTree` is deleted before the `TFile`. (If you do the reverse the deletion of the `TTree` will fails because .. it is already deleted).

Reply:
Applied unique_prt to all, i.e., TFile, TTree and TBranch, but did not consider order of deletion (in fact, no explicit deletion).

Reply:
The order of deletion is the inversion order of declaration.  So you need:
```
std::unique_ptr<TFile> file_;
std::unqiue_ptr<TTree> tree_;
```

Reply:
Good to know. Yes, it was in that order.

Reply:
Then it is a 'more subtle' problem.  To track it down, run
```
valgrind --log-file=myexec.01.vallog --suppression=$ROOTSYS/src/valgrind-root.supp myexecutable myargs
```

Reply:
==134133== Invalid free() / delete / delete[] / realloc()
==134133==    at 0x4C2B40D: operator delete(void*) (vg_replace_malloc.c:586)
==134133==    by 0x4280AF: operator() (unique_ptr.h:81)
==134133==    by 0x4280AF: ~unique_ptr (unique_ptr.h:292)
==134133==    by 0x4280AF: ~HitRootIO (HitRootIO.hh:53)
==134133==    by 0x4280AF: G4ThreadLocalSingleton<demo_geant::HitRootIO>::Clear() (G4ThreadLocalSingleton.hh:239)
==134133==    by 0x428187: G4ThreadLocalSingleton<demo_geant::HitRootIO>::~G4ThreadLocalSingleton() (G4ThreadLocalSingleton.hh:205)
==134133==    by 0xCD93CE8: __run_exit_handlers (in /usr/lib64/libc-2.17.so)

Which is consistent with the back tract from gdb that I sent to @stognini 
#0  0x00007fffcf8e9810 in ?? ()
#1  0x0000000000422e33 in operator() (this=0x7fffcd1fc378, 
    __ptr=<optimized out>)
    at /work1/celeritas/syjun/celeritas/src/celeritas/app/demo-geant-integration/HitRootIO.cc:66
#2  ~unique_ptr (this=0x7fffcd1fc378, __in_chrg=<optimized out>)
    at /srv/software/gnu9/9.3.0/include/c++/9.3.0/bits/unique_ptr.h:292
#3  demo_geant::HitRootIO::~HitRootIO (this=0x7fffcd1fc340, 
    __in_chrg=<optimized out>)
    at /work1/celeritas/syjun/celeritas/src/celeritas/app/demo-geant-integration/HitRootIO.cc:64
#4  0x0000000000428018 in G4ThreadLocalSingleton<demo_geant::HitRootIO>::Clear
    (this=this@entry=0x451320 <demo_geant::HitRootIO::GetInstance()::instance>)
    at /work1/celeritas/syjun/celeritas/src/geant4.11/source/global/management/include/G4ThreadLocalSingleton.hh:235
#5  0x00000000004280d8 in G4ThreadLocalSingleton<demo_geant::HitRootIO>::~G4ThreadLocalSingleton (
    this=0x451320 <demo_geant::HitRootIO::GetInstance()::instance>, 
    __in_chrg=<optimized out>)
    at /work1/celeritas/syjun/celeritas/src/geant4.11/source/global/management/include/G4ThreadLocalSingleton.hh:203
#6  0x00007fffefb22ce9 in __run_exit_handlers () from /lib64/libc.so.6

Reply:
Is that all the information in the log file?  Usually valgrind explains why the free is invalid.

Reply:
can you take a look at /work1/celeritas/syjun/celeritas/build/celeritas/app/simple-cms-test.vallog ?  Thanks.

Reply:
I had a typo instead of `--suppression=$ROOTSYS/src/valgrind-root.supp`
use `--suppression=$ROOTSYS/etc/valgrind-root.supp`

Reply:
The first problem I see is:
```
==134133== Invalid read of size 4
==134133==    at 0xC090264: G4NavigationLevel::~G4NavigationLevel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0xC08FA8B: G4NavigationHistoryPool::Clean() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0xC08FAF8: G4NavigationHistoryPool::~G4NavigationHistoryPool() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0x9A8EC1E: G4RunManagerKernel::~G4RunManagerKernel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A957F8: G4WorkerRunManagerKernel::~G4WorkerRunManagerKernel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A7C3D6: G4RunManager::~G4RunManager() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A889C8: G4WorkerRunManager::~G4WorkerRunManager() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A95590: G4MTRunManagerKernel::StartThread(G4WorkerThread*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0xC5336BF: execute_native_thread_routine (thread.cc:80)
==134133==    by 0x8AEBEA4: start_thread (in /usr/lib64/libpthread-2.17.so)
==134133==  Address 0x18b2ed58 is 472 bytes inside a block of size 1,008 free'd
==134133==    at 0x4C2BA7F: operator delete[](void*) (vg_replace_malloc.c:651)
==134133==    by 0xC2FAB5C: G4AllocatorPool::Reset() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4global.so)
==134133==    by 0xC2FAE73: G4AllocatorList::Destroy(int, int) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4global.so)
==134133==    by 0x9A8EC50: G4RunManagerKernel::~G4RunManagerKernel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A957F8: G4WorkerRunManagerKernel::~G4WorkerRunManagerKernel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A7C3D6: G4RunManager::~G4RunManager() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A889C8: G4WorkerRunManager::~G4WorkerRunManager() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A95590: G4MTRunManagerKernel::StartThread(G4WorkerThread*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0xC5336BF: execute_native_thread_routine (thread.cc:80)
==134133==    by 0x8AEBEA4: start_thread (in /usr/lib64/libpthread-2.17.so)
==134133==  Block was alloc'd at
==134133==    at 0x4C2AB28: operator new[](unsigned long) (vg_replace_malloc.c:433)
==134133==    by 0xC2FABBD: G4AllocatorPool::Grow() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4global.so)
==134133==    by 0xC090238: G4NavigationLevel::G4NavigationLevel() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0xC08F757: G4NavigationHistory::G4NavigationHistory(G4NavigationHistory const&) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0xC09AE12: G4TouchableHistory::G4TouchableHistory(G4NavigationHistory const&) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4geometry.so)
==134133==    by 0xABC9116: G4Transportation::PostStepDoIt(G4Track const&, G4Step const&) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4processes.so)
==134133==    by 0x9B84D1D: G4SteppingManager::InvokePSDIP(unsigned long) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4tracking.so)
==134133==    by 0x9B854CA: G4SteppingManager::InvokePostStepDoItProcs() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4tracking.so)
==134133==    by 0x9B81B2B: G4SteppingManager::Stepping() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4tracking.so)
==134133==    by 0x9B97047: G4TrackingManager::ProcessOneTrack(G4Track*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4tracking.so)
==134133==    by 0x9AEFC62: G4EventManager::DoProcessing(G4Event*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4event.so)
==134133==    by 0x9A887BB: G4WorkerRunManager::DoEventLoop(int, char const*, int) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
```

Reply:
The first ROOT related problem I see is:
```
==134133== Invalid read of size 8
==134133==    at 0x42808A: operator() (unique_ptr.h:81)
==134133==    by 0x42808A: ~unique_ptr (unique_ptr.h:292)
==134133==    by 0x42808A: ~HitRootIO (HitRootIO.hh:53)
==134133==    by 0x42808A: G4ThreadLocalSingleton<demo_geant::HitRootIO>::Clear() (G4ThreadLocalSingleton.hh:239)
==134133==    by 0x428187: G4ThreadLocalSingleton<demo_geant::HitRootIO>::~G4ThreadLocalSingleton() (G4ThreadLocalSingleton.hh:205)
==134133==    by 0xCD93CE8: __run_exit_handlers (in /usr/lib64/libc-2.17.so)
==134133==    by 0xCD93D36: exit (in /usr/lib64/libc-2.17.so)
==134133==    by 0xCD7C55B: (below main) (in /usr/lib64/libc-2.17.so)
==134133==  Address 0x17d5ecf0 is 0 bytes inside a block of size 1,040 free'd
==134133==    at 0x4C2B40D: operator delete(void*) (vg_replace_malloc.c:586)
==134133==    by 0x5F4242B: TObjArray::Delete(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==    by 0x413C09F: TTree::~TTree() (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libTree.so)
==134133==    by 0x413C688: TTree::~TTree() (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libTree.so)
==134133==    by 0x5F3C70F: TList::Delete(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==    by 0x5F32ABA: THashList::Delete(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==    by 0x59D71B4: TDirectoryFile::Close(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libRIO.so)
==134133==    by 0x59F44AB: TFile::Close(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libRIO.so)
==134133==    by 0x59F49B1: TFile::~TFile() (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libRIO.so)
==134133==    by 0x59F4C88: TFile::~TFile() (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libRIO.so)
==134133==    by 0x5F3C137: TList::Delete(char const*) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==    by 0x5E847A5: TROOT::~TROOT() (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==  Block was alloc'd at
==134133==    at 0x4C2A483: operator new(unsigned long) (vg_replace_malloc.c:344)
==134133==    by 0x5EDD399: TStorage::ObjectAlloc(unsigned long) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libCore.so)
==134133==    by 0x4141842: TTree::BronchExec(char const*, char const*, void*, bool, int, int) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libTree.so)
==134133==    by 0x412EEEF: TTree::BranchImp(char const*, TClass*, void*, int, int) (in /work1/celeritas/syjun/celeritas/install/v6-24-02/lib/libTree.so)
==134133==    by 0x422EF9: Branch<demo_geant::HitRootEvent> (TTree.h:368)
==134133==    by 0x422EF9: demo_geant::HitRootIO::WriteObject(demo_geant::HitRootEvent*) (HitRootIO.cc:127)
==134133==    by 0x423B25: demo_geant::HitRootIO::WriteHits(G4Event const*) (HitRootIO.cc:109)
==134133==    by 0x4192FF: demo_geant::EventAction::EndOfEventAction(G4Event const*) (EventAction.cc:54)
==134133==    by 0x9AEFAE3: G4EventManager::DoProcessing(G4Event*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4event.so)
==134133==    by 0x9A887BB: G4WorkerRunManager::DoEventLoop(int, char const*, int) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A7A8DD: G4RunManager::BeamOn(int, char const*, int) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A8BC1B: G4WorkerRunManager::DoWork() (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
==134133==    by 0x9A954C9: G4MTRunManagerKernel::StartThread(G4WorkerThread*) (in /work1/celeritas/syjun/celeritas/build/geant4.11/BuildProducts/lib/libG4run.so)
```

Reply:
And what is says is ... that static variable are evil ⟨™⟩ .   The problem is actually that the ROOT main static object is being tear down before the `G4ThreadLocalSingleton` object is tear down.  One of the job of the TROOT destructor is to close and delete the file (as anyway doing anything with them afterward would non-sensical).

The safest fix is to insure that all the `TFile` are closed and/or deleted *before* the end of `main` function.

Reply:
Okay the suggested fix resolved the segmentation fault when using unique_ptrs for TFile/TTree/TBranch.  Thanks!

Reply:
Did you also address the issue with `G4AllocatorPool`? (eg. re-run valgrind to see if all the reports are gone).

Reply:
The issue with G4AllocatorPool is still there, which may be nothing to do with this MR.

Reply:
@whokion I don't see any mention of an issue with `G4AllocatorPool`. Do you have a backtrace or link to it? @amandalund and I both have seen similar which we tracked down to an issue where Geant4 creates thread-local allocators and assumes that you always delete from the same thread you create.

Reply:
It is a few message back: https://github.com/celeritas-project/celeritas/pull/606#discussion_r1084683073
i.e. valgrind complains nicely.